pkg/streamingpromql/engine_test.go

```diff
@@ -1682,3 +1702,133 @@ func TestAnnotations(t *testing.T) {
+
+	pointsPerSeries := 6
+	samples := `
+		series{label="a", group="a"} 1 2 3 4 5 -50
+		series{label="b", group="a"} {{schema:1 sum:15 count:10 buckets:[3 2 5 7 9]}} {{schema:2 sum:20 count:15 buckets:[4]}} {{schema:3 sum:25 count:20 buckets:[5 8]}} {{schema:4 sum:30 count:25 buckets:[6 9 10 11]}} {{schema:5 sum:35 count:30 buckets:[7 10 13]}} {{schema:6 sum:40 count:35 buckets:[8 11 14]}}
+		series{label="c", group="a"} 1 {{schema:3 sum:5 count:3 buckets:[1 1 1]}} 3 {{schema:3 sum:10 count:6 buckets:[2 2 2]}} NaN {{schema:3 sum:12 count:7 buckets:[2 2 3]}}
+		series{label="d", group="a"} 1 _ _ stale 5 6
+		series{label="e", group="b"} {{schema:4 sum:12 count:8 buckets:[2 3 3]}} {{schema:4 sum:14 count:9 buckets:[3 3 3]}} _ stale {{schema:4 sum:18 count:11 buckets:[4 4 3]}} NaN
+		series{label="f", group="b"} 1 NaN _ 4 5 NaN
+		series{label="g", group="b"} NaN NaN NaN NaN NaN NaN
+		series{label="h", group="b"} NaN NaN NaN _ NaN stale
+		series{label="i", group="c"} 1 {{schema:5 sum:15 count:10 buckets:[3 2 5]}} _ NaN {{schema:2 sum:30 count:25 buckets:[6 9 10 9 1]}} stale
+		series{label="j", group="c"} 1 -20 stale stale stale stale
+		series{label="k", group="c"} 0 0 0 NaN stale 0
+		series{label="l", group="d"} {{schema:1 sum:10 count:5 buckets:[1 2]}} _ 3 _ stale NaN
+		series{label="m", group="d"} stale stale NaN _ _ 4
+		series{label="n", group="d"}
+	`
+
+	// Labels for generating combinations
+	labels := []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k"}
+
+	// Generate combinations of 2 and 3 labels. (e.g., "a,b", "e,f", "c,d,e" etc)
+	// These will be used for binary operations, so we can add up to 3 series.
+	labelCombinations := combinations(labels, 2)
+	labelCombinations = append(labelCombinations, combinations(labels, 3)...)
+
+	expressions := []string{}
+
+	// Binary operations
+	for _, labels := range labelCombinations {
+		if len(labels) >= 2 {
+			for _, op := range []string{"+", "-", "*", "/"} {
+				binaryExpr := fmt.Sprintf(`series{label="%s"}`, labels[0])
+				for _, label := range labels[1:] {
+					binaryExpr += fmt.Sprintf(` %s series{label="%s"}`, op, label)
+				}
+				expressions = append(expressions, binaryExpr)
+			}
+		}
+	}
+
+	// For aggregations, also add combinations of 4 labels. (e.g., "a,b,c,d", "c,d,e,f" etc)
+	labelCombinations = append(labelCombinations, combinations(labels, 4)...)
+
+	for _, labels := range labelCombinations {
+		labelRegex := strings.Join(labels, "|")
+		// Aggregations
+		for _, aggFunc := range []string{"sum", "avg", "min", "max"} {
+			expressions = append(expressions, fmt.Sprintf(`%s(series{label=~"(%s)"})`, aggFunc, labelRegex))
+			expressions = append(expressions, fmt.Sprintf(`%s by (group) (series{label=~"(%s)"})`, aggFunc, labelRegex))
+			expressions = append(expressions, fmt.Sprintf(`%s without (group) (series{label=~"(%s)"})`, aggFunc, labelRegex))
+		}
+		expressions = append(expressions, fmt.Sprintf(`rate(series{label=~"(%s)"}[45s])`, labelRegex))
```

Comment:
Why 45s? Doesn't this mean `rate` will always see at most one input sample per output sample and so never produce any output samples?

Reply:
Yes. This is an edge case I want to test. It can also affect the lookback etc.

Reply:
Might be good to add a comment explaining that, and would be good to add a case that does select multiple input samples as well.